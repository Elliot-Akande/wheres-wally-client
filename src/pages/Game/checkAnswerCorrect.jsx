const checkAnswerCorrect = (levelNum, answer) => {
  return true;
};

export default checkAnswerCorrect;
