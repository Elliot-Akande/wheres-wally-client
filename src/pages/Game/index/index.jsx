import { useEffect, useState } from "react";
import { Link, useParams } from "react-router-dom";
import useFetch from "../../../hooks/useFetch";
import fetchAsync from "../../../utils/fetchAsync";
import Checklist from "../Checklist/Checklist";
import LevelCompleteMenu from "../LevelCompleteMenu/LevelCompleteMenu";
import TaggableImage from "../TaggableImage/TaggableImage";

const Game = () => {
  const { levelNum } = useParams();
  const { data, loading, error } = useFetch(`/levels/${levelNum}`);

  const [correctAnswers, setCorrectAnswers] = useState([]);
  const [completionData, setCompletionData] = useState(null);

  useEffect(() => {
    const checkLevelComplete = async () => {
      const response = await fetchAsync(`/levels/${levelNum}/check-complete`, {
        method: "POST",
        headers: {
          Accept: "application/json",
          Authorization: `Bearer ${data.token}`,
          "Content-Type": "application/json",
        },
        body: JSON.stringify({ answers: correctAnswers }),
      });

      if (!(response instanceof Error) && response.isComplete) {
        setCompletionData(response);
      }
    };

    if (data !== null && correctAnswers.length === data.characters.length) {
      checkLevelComplete();
    }
  }, [correctAnswers]);

  const checkAnswer = async (answer) => {
    const data = await fetchAsync(`/levels/${levelNum}/check-answer`, {
      method: "POST",
      headers: {
        Accept: "application/json",
        "Content-Type": "application/json",
      },
      body: JSON.stringify(answer),
    });

    if (!(data instanceof Error) && data.isCorrect) {
      setCorrectAnswers((pastAnswers) => [...pastAnswers, answer]);
      return true;
    }

    return false;
  };

  if (loading) return <p>Loading...</p>;
  if (error) return <p>Error loading data.</p>;

  const remainingCharacters = data.characters.filter(
    (character) =>
      !correctAnswers.find((answer) => character.name === answer.character)
  );

  return (
    <>
      <h1>Level {levelNum}</h1>
      <TaggableImage
        imageUrl={data.imageUrl}
        checkAnswer={checkAnswer}
        characters={remainingCharacters}
        correctAnswers={correctAnswers}
        levelComplete={completionData !== null}
      />
      <Checklist characters={data.characters} correctAnswers={correctAnswers} />
      <Link to="/">Quit level</Link>

      {completionData !== null && (
        <LevelCompleteMenu
          levelNum={levelNum}
          token={completionData.token}
          score={completionData.score}
        />
      )}
    </>
  );
};

export default Game;
