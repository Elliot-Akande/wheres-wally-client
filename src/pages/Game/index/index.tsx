import { useEffect, useState } from "react";
import { useParams } from "react-router-dom";
import useFetch from "../../../hooks/useFetch";
import fetchAsync from "../../../utils/fetchAsync";
import Checklist from "../Checklist/Checklist";
import LevelCompleteMenu from "../LevelCompleteMenu/LevelCompleteMenu";
import TaggableImage from "../TaggableImage/TaggableImage";
import { TopContainer, Title } from "./styles";

interface Character {
  name: string;
  imageUrl: string;
}

interface Answer {
  xCoord: number;
  yCoord: number;
  character: string;
}

interface CorrectAnswer extends Answer {
  imageUrl: string;
}

interface Data {
  token: string;
  imageUrl: string;
  characters: Character[];
}

interface CompletionData {
  score: number;
  token: string;
  isComplete: boolean;
}

const Game = () => {
  const { levelNum } = useParams();
  const { data, loading, error } = useFetch<Data>(`/levels/${levelNum}`);

  const [correctAnswers, setCorrectAnswers] = useState<CorrectAnswer[]>([]);
  const [completionData, setCompletionData] = useState<CompletionData | null>(
    null
  );

  useEffect(() => {
    const checkLevelComplete = async (): Promise<void> => {
      const response = await fetchAsync<CompletionData>(
        `/levels/${levelNum}/check-complete`,
        {
          method: "POST",
          headers: {
            Accept: "application/json",
            Authorization: `Bearer ${data!.token}`,
            "Content-Type": "application/json",
          },
          body: JSON.stringify({ answers: correctAnswers }),
        }
      );

      if (!(response instanceof Error) && response.isComplete) {
        setCompletionData(response);
      }
    };

    if (data && correctAnswers.length === data.characters.length) {
      checkLevelComplete();
    }
  }, [correctAnswers]);

  if (loading) return <p>Loading...</p>;
  if (error || !data) return <p>Error loading data.</p>;

  const checkAnswer = async (answer: Answer): Promise<boolean> => {
    const response = await fetchAsync<{ isCorrect: boolean }>(
      `/levels/${levelNum}/check-answer`,
      {
        method: "POST",
        headers: {
          Accept: "application/json",
          "Content-Type": "application/json",
        },
        body: JSON.stringify(answer),
      }
    );

    if (!(response instanceof Error) && response.isCorrect) {
      // Add imageUrl for displaying correct answers on game image.
      const imageUrl = data.characters.find(
        (character) => character.name === answer.character
      )!.imageUrl;

      setCorrectAnswers((pastAnswers) => [
        ...pastAnswers,
        { ...answer, imageUrl },
      ]);
      return true;
    }

    return false;
  };

  const remainingCharacters = data.characters.filter(
    (character) =>
      !correctAnswers.find((answer) => character.name === answer.character)
  );

  return (
    <>
      <TopContainer>
        <Title>LEVEL {levelNum}</Title>
        <Checklist
          characters={data.characters}
          correctAnswers={correctAnswers}
        />
      </TopContainer>
      <TaggableImage
        imageUrl={data.imageUrl}
        checkAnswer={checkAnswer}
        characters={remainingCharacters}
        correctAnswers={correctAnswers}
        levelComplete={completionData !== null}
      />

      {completionData !== null && (
        <LevelCompleteMenu
          levelNum={levelNum}
          token={completionData.token}
          score={completionData.score}
        />
      )}
    </>
  );
};

export default Game;
