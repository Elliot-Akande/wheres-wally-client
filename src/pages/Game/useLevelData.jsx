const useLevelData = (levelNum) => {
  return {
    data: {
      token: "token",
      img: "#",
      characters: ["Wally", "Woof", "Odlaw"],
    },
    loading: false,
    error: null,
  };
};

export default useLevelData;
