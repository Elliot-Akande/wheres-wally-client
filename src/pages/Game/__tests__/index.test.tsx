import { render, screen } from "@testing-library/react";
import { userEvent } from "@testing-library/user-event";
import { MemoryRouter } from "react-router-dom";
import { describe, expect, it, vi } from "vitest";
import Game from "../index";

const mocks = vi.hoisted(() => ({
  useParams: vi.fn(),
  fetchAsync: vi.fn(),
  useFetch: vi.fn(),
}));

vi.mock("react-router-dom", async () => {
  const mod = await vi.importActual("react-router-dom");
  return {
    ...mod,
    useParams: mocks.useParams,
  };
});
vi.mock("../../../utils/fetchAsync.js", () => ({
  default: mocks.fetchAsync,
}));
vi.mock("../../../hooks/useFetch.jsx", () => ({
  default: mocks.useFetch,
}));

vi.mock("../Checklist/Checklist.jsx", () => ({
  default: () => <div>Checklist Component</div>,
}));
vi.mock("../LevelCompleteMenu/LevelCompleteMenu.jsx", () => ({
  default: () => <div>LevelCompleteMenu Component</div>,
}));
vi.mock("../TaggableImage/TaggableImage.jsx", () => ({
  default: ({ checkAnswer }) => {
    return (
      <div>
        TaggableImage Component
        <button
          onClick={() =>
            checkAnswer({ xCoord: 10, yCoord: 10, character: "Wally" })
          }
        >
          Wally
        </button>
        <button
          onClick={() =>
            checkAnswer({ xCoord: 20, yCoord: 20, character: "Odlaw" })
          }
        >
          Odlaw
        </button>
      </div>
    );
  },
}));

describe("Game", () => {
  const levelNum = 3;
  const data = {
    token: "token",
    imageUrl: "/game.png",
    characters: [
      {
        name: "Wally",
        imageUrl: "/wally.png",
      },
      {
        name: "Odlaw",
        imageUrl: "/odlaw.png",
      },
    ],
  };

  it("renders heading, Checklist, and TaggableImage", () => {
    mocks.useParams.mockReturnValue({ levelNum });
    mocks.useFetch.mockReturnValue({ data, loading: false, error: null });

    const { container } = render(
      <MemoryRouter>
        <Game />
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
  });

  it("renders loading message when loading level data", () => {
    mocks.useParams.mockReturnValue({ levelNum });
    mocks.useFetch.mockReturnValue({ data: null, loading: true, error: null });

    const { container } = render(
      <MemoryRouter>
        <Game />
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
  });

  it("renders error message level data error", () => {
    mocks.useParams.mockReturnValue({ levelNum });
    mocks.useFetch.mockReturnValue({
      data,
      loading: false,
      error: new Error(),
    });

    const { container } = render(
      <MemoryRouter>
        <Game />
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
  });

  it("renders LevelCompleteMenu when all characters are found", async () => {
    const user = userEvent.setup();
    mocks.useParams.mockReturnValue({ levelNum });
    mocks.useFetch.mockReturnValue({ data, loading: false, error: null });
    mocks.fetchAsync
      .mockResolvedValueOnce({ isCorrect: true })
      .mockResolvedValueOnce({ isCorrect: true })
      .mockResolvedValueOnce({ isComplete: true });

    render(
      <MemoryRouter>
        <Game />
      </MemoryRouter>
    );

    const buttons = data.characters.map(({ name }) =>
      screen.getByRole("button", { name })
    );

    expect(screen.queryByText("LevelCompleteMenu Component")).toBeNull();

    await user.click(buttons[0]);
    await user.click(buttons[1]);
    expect(screen.getByText("LevelCompleteMenu Component")).toBeInTheDocument();
  });

  it("doesn't render levelCompleteMenu when some but not all characters are found", async () => {
    const user = userEvent.setup();
    mocks.useParams.mockReturnValue({ levelNum });
    mocks.useFetch.mockReturnValue({ data, loading: false, error: null });
    mocks.fetchAsync
      .mockResolvedValueOnce({ isCorrect: true })
      .mockResolvedValueOnce({ isComplete: false });

    render(
      <MemoryRouter>
        <Game />
      </MemoryRouter>
    );

    const buttons = data.characters.map(({ name }) =>
      screen.getByRole("button", { name })
    );

    await user.click(buttons[0]);

    expect(screen.queryByText("LevelCompleteMenu Component")).toBeNull();
  });
});
