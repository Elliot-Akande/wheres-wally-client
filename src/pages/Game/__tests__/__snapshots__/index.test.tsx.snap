// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Game > renders error message level data error 1`] = `
<div>
  <p>
    Error loading data.
  </p>
</div>
`;

exports[`Game > renders heading, Checklist, and TaggableImage 1`] = `
<div>
  <div
    class="sc-beySPh gLCnrv"
  >
    <h1
      class="sc-guDLey cxjOww"
    >
      LEVEL 
      3
    </h1>
    <div>
      Checklist Component
    </div>
  </div>
  <div>
    TaggableImage Component
    <button>
      Wally
    </button>
    <button>
      Odlaw
    </button>
  </div>
</div>
`;

exports[`Game > renders loading message when loading level data 1`] = `
<div>
  <p>
    Loading...
  </p>
</div>
`;
